File: paleo_workbench/providers/contracts.py
"""Capability provider contracts (P2-B).

A *capability provider* is the sanctioned extension unit of the workbench
(ADR 0055 track P.REG): a small object exposing a frozen
:class:`ProviderDescriptor` (structured metadata, never introspected Python
signatures) and an ``execute`` method over typed inputs/outputs whose data
outputs enter the catalog through :class:`paleo_workbench.catalog.port.CatalogPort`.

Design rules:

- **Descriptors are data**: provider_id/family/version, JSON-schema
  parameters, typed input/output names, resource profile, threading model.
  Stable, testable, JSON-serializable.
- **Typed inputs**: providers receive :class:`TypedRef` instances
  (``DataVersionRef``, ``WellRef``, ``SeismicVolumeRef``, …) — never random
  dicts to be guessed apart. Untyped payloads are rejected by the executor.
- **Outputs are results**: :class:`ProviderResult` with artifacts, warnings,
  diagnostics, provenance and metrics — one small object, not a god-class.
- **Families stay separate**: no mega ``BasePlugin``. The family only fixes
  vocabulary; behaviour contracts live with the deep seams each family wraps
  (CatalogPort, Interpolator, KERNELS, ModelProvider, MapRenderBackend…).
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

from paleo_workbench.providers.errors import InvalidProviderError

_ID_RE = re.compile(r"^[a-z0-9][a-z0-9._-]{1,63}$")
_VERSION_RE = re.compile(r"^[0-9]+(\.[0-9]+){0,3}$")


class ProviderFamily(str, Enum):
    """Families with real production consumers today.

    ``DATA_FORMAT`` / ``PREVIEW`` / ``MAP_COMPONENT`` exist in the vocabulary
    because the seams are known (FormatSpec/PreviewRegistry/composer), but no
    built-in providers ship for them yet — the registry accepts them so
    third parties can register without a schema change.
    """

    INTERPOLATION = "interpolation"
    SEISMIC_ATTRIBUTE = "seismic_attribute"
    INFERENCE = "inference"
    VISUALIZATION = "visualization"
    EXPORTER = "exporter"
    IMPORTER = "importer"
    DATA_FORMAT = "data_format"
    PREVIEW = "preview"
    MAP_COMPONENT = "map_component"


#: Typed reference vocabulary. Keys are the names usable in
#: ``input_types``/``output_types``; values document what concrete type the
#: executor resolves them to. ``DataVersionRef`` is the catalog's own DTO —
#: the artifact authority — so every data output can be traced.
TYPED_REFS: dict[str, str] = {
    "DataVersionRef": "paleo_workbench.catalog.types.DataVersionRef (catalog version identity)",
    "WellRef": "paleo_workbench.providers.refs.WellRef (project well identity)",
    "SeismicVolumeRef": "paleo_workbench.providers.refs.SeismicVolumeRef (volume path/store identity)",
    "MapDocumentRef": "paleo_workbench.providers.refs.MapDocumentRef (map document identity)",
    "FactorDatasetRef": "paleo_workbench.providers.refs.FactorDatasetRef (factor dataset handle)",
    "FactorGridRef": "paleo_workbench.providers.refs.FactorGridRef (interpolated grid handle)",
    "PathRef": "paleo_workbench.providers.refs.PathRef (workspace-relative file handle)",
    # In-process domain objects providers may receive directly (typed, never
    # anonymous dicts):
    "GeologicalFactorDataset": "paleo_workbench.mapping.geological_pipeline.models.GeologicalFactorDataset",
    "MapDocument": "paleo_workbench.mapping.layers.MapDocument",
}


@dataclass(frozen=True, slots=True)
class ResourceProfile:
    """Declared resource cost of one provider execution (estimates, not
    pre-allocation) — consumed by the ResourceGovernor for admission."""

    estimated_cpu_cores: float = 1.0
    estimated_ram_bytes: int = 0
    estimated_vram_bytes: int = 0
    io_weight: float = 1.0
    category: str = "background.compute"  # TaskCategory value for admission

    def to_dict(self) -> dict[str, Any]:
        return {
            "estimated_cpu_cores": self.estimated_cpu_cores,
            "estimated_ram_bytes": self.estimated_ram_bytes,
            "estimated_vram_bytes": self.estimated_vram_bytes,
            "io_weight": self.io_weight,
            "category": self.category,
        }


@dataclass(frozen=True, slots=True)
class ProviderDescriptor:
    """Structured, stable metadata for one provider (never guessed)."""

    provider_id: str
    family: ProviderFamily
    version: str
    display_name: str
    description: str = ""
    capabilities: tuple[str, ...] = ()
    input_types: tuple[str, ...] = ()
    output_types: tuple[str, ...] = ()
    parameters_schema: dict[str, Any] = field(default_factory=dict)
    resource_profile: ResourceProfile = field(default_factory=ResourceProfile)
    supports_cancel: bool = False
    supports_resume: bool = False
    deterministic: bool = True
    threading_model: str = "worker_thread"  # worker_thread | gui_thread | any

    def to_dict(self) -> dict[str, Any]:
        return {
            "provider_id": self.provider_id,
            "family": self.family.value,
            "version": self.version,
            "display_name": self.display_name,
            "description": self.description,
            "capabilities": list(self.capabilities),
            "input_types": list(self.input_types),
            "output_types": list(self.output_types),
            "parameters_schema": self.parameters_schema,
            "resource_profile": self.resource_profile.to_dict(),
            "supports_cancel": self.supports_cancel,
            "supports_resume": self.supports_resume,
            "deterministic": self.deterministic,
            "threading_model": self.threading_model,
        }


def validate_descriptor(descriptor: ProviderDescriptor) -> list[str]:
    """Structural validation; returns a list of problems (empty = valid)."""
    problems: list[str] = []
    if not _ID_RE.match(descriptor.provider_id or ""):
        problems.append(f"provider_id {descriptor.provider_id!r} must match {_ID_RE.pattern}")
    if not _VERSION_RE.match(descriptor.version or ""):
        problems.append(f"version {descriptor.version!r} must be numeric dotted (e.g. 1.0.0)")
    if not isinstance(descriptor.family, ProviderFamily):
        problems.append(f"family {descriptor.family!r} is not a ProviderFamily")
    if not (descriptor.display_name or "").strip():
        problems.append("display_name must be a non-empty string")
    schema = descriptor.parameters_schema
    if not isinstance(schema, dict):
        problems.append("parameters_schema must be a dict (JSON schema)")
    else:
        if schema.get("type") not in (None, "object"):
            problems.append("parameters_schema must describe an object at the top level")
        props = schema.get("properties")
        if props is not None and not isinstance(props, dict):
            problems.append("parameters_schema.properties must be a dict")
        required = schema.get("required")
        if required is not None and not isinstance(required, list):
            problems.append("parameters_schema.required must be a list")
    for role, types in (("input_types", descriptor.input_types), ("output_types", descriptor.output_types)):
        for t in types:
            if t not in TYPED_REFS:
                problems.append(f"{role} entry {t!r} is not a known typed ref {sorted(TYPED_REFS)}")
    if descriptor.threading_model not in ("worker_thread", "gui_thread", "any"):
        problems.append(f"threading_model {descriptor.threading_model!r} invalid")
    profile = descriptor.resource_profile
    if not isinstance(profile, ResourceProfile):
        problems.append("resource_profile must be a ResourceProfile")
    else:
        if profile.estimated_cpu_cores <= 0:
            problems.append("resource_profile.estimated_cpu_cores must be > 0")
        if profile.io_weight < 0:
            problems.append("resource_profile.io_weight must be >= 0")
    return problems


def assert_valid_descriptor(descriptor: ProviderDescriptor) -> None:
    problems = validate_descriptor(descriptor)
    if problems:
        raise InvalidProviderError(descriptor.provider_id, problems)

File: libs/providers/src/contracts.cpp
#include <pwb/providers/contracts.hpp>

#include <pwb/providers/errors.hpp>

#include <algorithm>
#include <cctype>
#include <regex>

namespace pwb::providers {

namespace {
// Python contracts.py: ^[a-z0-9][a-z0-9._-]{1,63}$ / ^[0-9]+(\.[0-9]+){0,3}$
const std::regex& id_pattern() {
    static const std::regex pattern(R"(^[a-z0-9][a-z0-9._-]{1,63}$)");
    return pattern;
}
const std::regex& version_pattern() {
    static const std::regex pattern(R"(^[0-9]+(\.[0-9]+){0,3}$)");
    return pattern;
}

// Python repr of a string: single quotes preferred, double quotes when the
// value contains a single quote but no double quote, and an escaped single
// quote when both kinds appear.
std::string repr_string(const std::string& value) {
    const bool has_single = value.find('\'') != std::string::npos;
    const bool has_double = value.find('"') != std::string::npos;
    if (has_single && !has_double) {
        return "\"" + value + "\"";
    }
    if (has_single) {
        std::string out = "'";
        for (char c : value) {
            if (c == '\'') out += "\\'";
            else out += c;
        }
        return out + "'";
    }
    return "'" + value + "'";
}

// Python repr of a list of strings: ['a', 'b'].
std::string repr_string_list(const std::vector<std::string>& values) {
    std::string out = "[";
    for (std::size_t i = 0; i < values.size(); ++i) {
        if (i != 0) out += ", ";
        out += repr_string(values[i]);
    }
    return out + "]";
}
}  // namespace

std::string to_string(ProviderFamily family) {
    switch (family) {
        case ProviderFamily::Interpolation: return "interpolation";
        case ProviderFamily::SeismicAttribute: return "seismic_attribute";
        case ProviderFamily::Inference: return "inference";
        case ProviderFamily::Visualization: return "visualization";
        case ProviderFamily::Exporter: return "exporter";
        case ProviderFamily::Importer: return "importer";
        case ProviderFamily::DataFormat: return "data_format";
        case ProviderFamily::Preview: return "preview";
        case ProviderFamily::MapComponent: return "map_component";
    }
    return "interpolation";
}

std::optional<ProviderFamily> family_from_string(std::string_view value) {
    if (value == "interpolation") return ProviderFamily::Interpolation;
    if (value == "seismic_attribute") return ProviderFamily::SeismicAttribute;
    if (value == "inference") return ProviderFamily::Inference;
    if (value == "visualization") return ProviderFamily::Visualization;
    if (value == "exporter") return ProviderFamily::Exporter;
    if (value == "importer") return ProviderFamily::Importer;
    if (value == "data_format") return ProviderFamily::DataFormat;
    if (value == "preview") return ProviderFamily::Preview;
    if (value == "map_component") return ProviderFamily::MapComponent;
    return std::nullopt;
}

const std::vector<std::string>& typed_refs() {
    static const std::vector<std::string> kRefs = {
        "DataVersionRef", "FactorDatasetRef", "FactorGridRef",
        "GeologicalFactorDataset", "MapDocument", "MapDocumentRef",
        "PathRef", "SeismicVolumeRef", "WellRef",
    };
    return kRefs;
}

Json ResourceProfile::to_json() const {
    Json out = Json::object();
    out["estimated_cpu_cores"] = estimated_cpu_cores;
    out["estimated_ram_bytes"] = estimated_ram_bytes;
    out["estimated_vram_bytes"] = estimated_vram_bytes;
    out["io_weight"] = io_weight;
    out["category"] = category;
    return out;
}

Json ProviderDescriptor::to_json() const {
    Json out = Json::object();
    out["provider_id"] = provider_id;
    out["family"] = to_string(family);
    out["version"] = version;
    out["display_name"] = display_name;
    out["description"] = description;
    out["capabilities"] = capabilities;
    out["input_types"] = input_types;
    out["output_types"] = output_types;
    out["parameters_schema"] = parameters_schema;
    out["resource_profile"] = resource_profile.to_json();
    out["supports_cancel"] = supports_cancel;
    out["supports_resume"] = supports_resume;
    out["deterministic"] = deterministic;
    out["threading_model"] = threading_model;
    if (build_identity.has_value()) {
        out["build_identity"] = *build_identity;
    } else {
        out["build_identity"] = Json(nullptr);
    }
    return out;
}

std::vector<std::string> validate_descriptor(const ProviderDescriptor& descriptor) {
    std::vector<std::string> problems;

    if (!std::regex_match(descriptor.provider_id, id_pattern())) {
        problems.push_back("provider_id " + repr_string(descriptor.provider_id) +
                           " must match ^[a-z0-9][a-z0-9._-]{1,63}$");
    }
    if (!std::regex_match(descriptor.version, version_pattern())) {
        problems.push_back("version " + repr_string(descriptor.version) +
                           " must be numeric dotted (e.g. 1.0.0)");
    }
    // family: a C++ enum value is always a ProviderFamily (compile-time).
    const bool display_name_blank = std::all_of(descriptor.display_name.begin(),
                                                descriptor.display_name.end(),
                                                [](unsigned char c) {
                                                    return std::isspace(c) != 0;
                                                });
    if (display_name_blank) {
        problems.push_back("display_name must be a non-empty string");
    }

    const Json& schema = descriptor.parameters_schema;
    if (!schema.is_object()) {
        problems.push_back("parameters_schema must be a dict (JSON schema)");
    } else {
        const Json schema_type = schema.contains("type") ? schema.at("type") : Json(nullptr);
        if (!(schema_type.is_null() || (schema_type.is_string() && schema_type == "object"))) {
            problems.push_back("parameters_schema must describe an object at the top level");
        }
        if (schema.contains("properties") && !schema.at("properties").is_object()) {
            problems.push_back("parameters_schema.properties must be a dict");
        }
        if (schema.contains("required") && !schema.at("required").is_array()) {
            problems.push_back("parameters_schema.required must be a list");
        }
    }

    const auto check_typed_refs = [&problems](const std::string& role,
                                              const std::vector<std::string>& types) {
        for (const auto& type : types) {
            if (std::find(typed_refs().begin(), typed_refs().end(), type) ==
                typed_refs().end()) {
                problems.push_back(role + " entry " + repr_string(type) +
                                   " is not a known typed ref " +
                                   repr_string_list(typed_refs()));
            }
        }
    };
    check_typed_refs("input_types", descriptor.input_types);
    check_typed_refs("output_types", descriptor.output_types);

    if (descriptor.threading_model != "worker_thread" &&
        descriptor.threading_model != "gui_thread" && descriptor.threading_model != "any") {
        problems.push_back("threading_model " + repr_string(descriptor.threading_model) +
                           " invalid");
    }
    // build_identity: std::optional<std::string> is always a string or null.

    if (descriptor.resource_profile.estimated_cpu_cores <= 0.0) {
        problems.push_back("resource_profile.estimated_cpu_cores must be > 0");
    }
    if (descriptor.resource_profile.io_weight < 0.0) {
        problems.push_back("resource_profile.io_weight must be >= 0");
    }
    return problems;
}

void assert_valid_descriptor(const ProviderDescriptor& descriptor) {
    auto problems = validate_descriptor(descriptor);
    if (!problems.empty()) {
        throw InvalidProviderError(descriptor.provider_id, std::move(problems));
    }
}

}  // namespace pwb::providers
